### ISSUE TITLE
update dev docs to talk about Python 3

### ISSUE THREAD

**Issue by faassen** [morepath/morepath#473]
I'm still using buildout with Python 2 to run tests, and that's probably wrong now.

@href, @henri-hulski, @taschini am I correct that you all prefer a virtualenv solution above buildout? If so, I think we should look into switching the dev environment away from buildout. But then we need to update ``developing.rst`` to talk about how to set up a virtualenv, how to run the tests, how to run the doctests, how to build the docs, how to do coverage, and how to do other checks such as radon. I don't even know how to do many things without buildout so it would help *me*.

Once it's okay for Morepath we should also update the development docs in Dectate and Reg.

**Comment by href** [comment]
I'm personally okay with either. I tend to use buildout for god-projects manage many different repositories, virtualenv for simple modules.

One thing I often missed with non-buildout environments was [collective.recipe.omelette](https://pypi.python.org/pypi/collective.recipe.omelette/), which creates a bunch of symlinks for easy module access in my editor. Luckily, there's a non-buildout alternative called [scrambler](https://pypi.python.org/pypi/collective.recipe.omelette/).

Another good thing with buildout is the ability to compile external dependencies like node, but Morepath doesn't have that.

**Comment by henri-hulski** [comment]
As we use Python 3 now we could also consider to recommend the [venv](https://docs.python.org/3/library/venv.html#module-venv)  module instead of virtualenv, like @blaflamme proposes in #456.

> @henri-hulski @faassen for python 3.5 I would opt for the venv module instead of virtualenv

> ``` python 
$ python3 -m venv myenv
```

**Comment by faassen** [comment]
I don't want to necessarily make life harder to work in Python 2: we still need to run the tests for Python 2 and debug them. Is running tox restricting the environment to Python 2 good enough for this?

We could explain how to do virtualenv and easily explain how to install stuff for Python 2. But with venv we lose that option. Does does venv bring to the table that virtualenv doesn't have?

**Comment by faassen** [comment]
Hm, at least in Reg in Python 3 I'm getting really weird line numbers when there are errors in doctests; nowhere near where the actual failure is. Can we see this with the Python 3 doctests of Morepath too?

**Comment by taschini** [comment]
(Sorry, pressed the wrong button)

The big advantage of ``buildout`` and one of its main selling points is the simplicity it provides when you document your project.  In a very few lines you're done (basically run bootstrap and then buildout).

I'd be tempted of leaving the docs as they are: if you need the docs to get started developing, buildout is probably your better option.  If you have strong reasons to use virtualenv, you probably already know how to use it.

**Comment by faassen** [comment]
Okay, if we're okay with buildout I'm fine continuing with it. We do need to add a note to buildout on how to run it with Python 3 - running the bootstrap with it.

**Comment by henri-hulski** [comment]
Actually I have no experience with venv, just wanted to mention the proposition of @blaflamme.
At the first glance it looks cleaner with a simpler directory structure and more flexible.
Maybe @blaflamme can give more details.

**Comment by henri-hulski** [comment]
The advantage of switching to virtualenv would be to have a consistent deployment flow all over the project.

**Comment by faassen** [comment]
Can you elaborate on "consistent deployment flow"?

**Comment by henri-hulski** [comment]
Just mean we would use similar tools in all apps of the project which makes it easier for newcomers to step in.

**Comment by blaflamme** [comment]
My point was that `python 3` provides `virtualenv` facility out of the box so no need for external dependencies and most people doing web, except from zope/plone world, would be familiar with this approach. Then I think it's easier for most newcomers to use and understand `virtualenv` than `buildout` because it's simpler and ton of examples already exist out there. And now a wider range of tools exist, like Docker, to help building environments and do better at separating concerns.

**Comment by taschini** [comment]
@faassen 
> We do need to add a note to buildout on how to run it with Python 3 - running the bootstrap with it.

That has already been done.

**Comment by faassen** [comment]
While it's true python 3 provides virtualenv we are going to support Python 2 for the forseeable future. At this point in time I'm only doing this Python 3 stuff for the community, not because it's being used for anything I do with Morepath. That may change, but Python 2 will be with us for many years to come.

So while I'm okay with switching to virtualenv for the purpose of familiarity, I'm not okay with switching to something available in Python 3.5 and further.

So the choice is between buildout and virtualenv. I'm leaning to sticking with buildout for the time being. as it's already automated and we've documented it and the current developers are okay with using it. But if someone is willing to do the work to document a virtualenv based environment that has all the features required, including things like building docs (possibly mixing tox in, but restricted to running tests and continuous integration), I'd be happy to evaluate it.

(I don't want to get into a big discussion but I don't accept the notion that docker is a useful tool to set up *development environments*; talk about learning overhead for developers!)

**Comment by henri-hulski** [comment]
After thinking about it a little more I also came to the conclusion that until we stop supporting Python 2 we should use virtualenv not venv.
Also agree with your opinion about docker.

**Comment by taschini** [comment]
@faassen I completely agree with what you wrote.  I suggest this way forward:

1. The current ``developing.rst`` with Buildout is basically fine as it is as Henry has already updated it to mention Python 3.5 instead of 2.7.  There are a few things related to Sphinxbuilder that  need to be updated, though — see below.

2. I suggest we create a second ``developing.rst`` which mimics the contents of the existing one but using Virtualenv instead of Buildout.  Then we can say: "You can setup a development environment for Morepath using Buildout or Virtualenv" with links to the two versions of  ``developing.rst``.

Introducing the Virtualenv option alongside Buildout is becoming necessary, I'm afraid.  I see that the ecosystem around Buildout is shrinking rapidly with many recipes that are bit-rotting due to their maintainers having switched to Virtualenv + Pip — if I'm not mistaken the Sphinxbuilder recipe has recently been taken over by Reinout just to rescue it from wasting away.   I also must make it clear that, while I don't mind having the docs present the Buildout option, I stopped using it, and so even within the core Morepath developers the know-how pertaining to Buildout is fading.

Even before we start with Point 2, there is some preparatory work that needs to done: currently, project dependencies are interspersed in three different places: ``buildout.cfg``, ``setup.py`` and ``tox.ini``. These need to be rationalized a bit, ideally so that the four sets of dependencies (install, testing, doc testing & generation, full development environment) are defined in one place and used by Buildout, Pip, and Tox as appropriate.

Getting back to  ``developing.rst`` as it is now,  what needs to be fixed is indeed related to Sphinxbuilder. In there we have

```rest
Building the HTML documentation
-------------------------------

To build the HTML documentation (output in ``doc/build/html``), run::

  $ bin/sphinxbuilder

```

If you look inside ``bin/sphinxbuilder`` you'll see something like

```sh
cd /some/path/morepath/doc/build
make html
```

so pretty much everything that is said in the preceding paragraph of ``developing.rst`` about having Make installed is a bit moot:  if you don't have Make, Sphinxbuilder won't help you much.  In fact, if you don't have Make, you might be better off using ``zc.recipe.egg`` with Sphinx instead of Sphinxbuilder.

In conclusion, I propose the following tasks:

1. Reorganize the dependencies between ``setup.py``, ``tox.ini`` and ``buildout.cfg``.
2. Amend the section of ``developing.rst`` related to doc testing and generation.
3. Clone ``developing.rst`` and adapt it to Virtualenv + Pip.

If you guys agree I can add them to the issue tracker.

**Comment by henri-hulski** [comment]
@taschini @faassen my opinion:

1. If we decide to switch to pip and virtualenv we should drop buildout entirely as it gets much cleaner and easier to maintain. I'm also a fan of having one way to do things.

2. I'm agree that it makes sense to have the dependencies on one place. We can keep them in setup.py and then create a test_requirements.txt file, which contains only `-e '.[test]'`, where the test extras in setup.py contains the dependencies we need for tox. In tox.ini we declare the dependencies like `deps = -rtest_requirements.txt`. Not sure if we can achieve something similar with the tox install_command setting.

**Comment by blaflamme** [comment]
Sorry for the confusion, it was not my intention to bring a discussion about using `docker`. My point was more about `buildout` can do a lot of things but I think there are more appropriated and used tools to achieve the same goal nowadays. If `buildout` is only used as a virtualenv, I think `virtualenv` is more appropriate, if `buildout` is used to build a full development env with all sort of external dependencies, `docker` is more appropriate imho. I think if a new comer has to learn something new it's better to learn the right tools.

I also think using `virtualenv` until we get rid of Python2 is the right thing, but I would probably update examples and tutorials to use Python3 with the included command. Developing projects with Morepath has no dependency on Python2 imho.

I also think Henri's idea about tox and tests would work just fine. Other projects have multiple requirements files structured as:

```
requirements/tests.txt
requirements/docs.txt
...
```

That could be a way of doing it, but best would be to find a way to just call extras somehow like `$ pip install -e -.[tests]`

**Comment by henri-hulski** [comment]
Sorry but my second point is already solved with `deps = -e{toxinidir}[test]`. 
So in tox.ini there are only  the dependencies flake8,  coveralls and sphinx left. 
Don't know why flake8 is not in test, sphinx could go in docs extras, but coveralls is actually only necessary for Travis. So no idea how to handle this.

**Comment by henri-hulski** [comment]
I just installed Python 3.3, 3.5, Pypy and Pypy3 with [pyenv](https://github.com/yyuu/pyenv) in addition to Python 2.7 and 3.4 which I have from the system.
After upgrading the systemwide virtualenv, setuptools and pip with
```sh
$ su
$ pip install -U virtualenv, setuptools pip
```
and installing tox `$ pip install -U tox`
I could run all tox test by just typing `tox` or selecting a specific one e.g. `tox -e pep8`.
First time it need some time to build the virtualenvs, but after it works fluently.

If we add `skip_missing_interpreters = True` to tox.ini the tests for python versions, which are not installed on the system would be skipped.
But the basepython for pep8, coverage and docs tests must be installed for running them.
This is for packages, which have already switched 3.5 and for the others 2.7

**Comment by henri-hulski** [comment]
Closed in #484.
